Src/GoogleApis.Core/Apis/Util/PageStreamer.cs

```diff
@@ -0,0 +1,137 @@
+            }
+            if (resourceExtractor == null)
+            {
+                throw new ArgumentNullException("resourceExtractor");
+            }
+            this._requestProvider = requestProvider;
+            this._tokenExtractor = tokenExtractor;
+            this._resourceExtractor = resourceExtractor;
+            this._emptyToken = emptyToken;
+        }
+
+        /// <summary>
+        /// Lazily fetches resources a page at a time.
+        /// </summary>
+        /// <param name="initialRequest">The initial request to send. If this contains a page token,
+        /// that token is maintained. This is also passed to the request provider along with a token
+        /// to provide another request. (The provider may build a new object, or modify the existing request.)</param>
+        /// <param name="responseFetcher">A function to fetch a response containing a page of results given a request.</param>
+        /// <returns>A sequence of resources, which are fetched a page at a time. Must not be null.</returns>
+        public IEnumerable<TResource> Fetch(TRequest initialRequest, Func<TRequest, TResponse> responseFetcher)
+        {
+            if (responseFetcher == null)
+            {
+                throw new ArgumentNullException("responseFetcher");
+            }
+            TToken token;
+            TRequest nextRequest = initialRequest;
+            do
+            {
+                TResponse response = responseFetcher(nextRequest);
+                token = _tokenExtractor(response);
+                nextRequest = _requestProvider(initialRequest, token);
+                foreach (var item in _resourceExtractor(response))
+                {
+                    yield return item;
+                }
+            } while (!IsEmptyToken(token));
+        }
+
+        /// <summary>
+        /// Asynchronously (but asynchronously) fetches a complete set of resources, potentially making multiple requests.
+        /// </summary>
+        /// <param name="initialRequest">The initial request to send. If this contains a page token,
+        /// that token is maintained. This is also passed to the request provider along with a token
+        /// to provide another request. (The provider may build a new object, or modify the existing request.)</param>
+        /// <param name="responseFetcher">A function to asynchronously fetch a response containing a
+        /// page of results given a request. Must not be null.</param>
+        /// <returns>A sequence of resources, which are fetched asynchronously and a page at a time.</returns>
+        /// <param name="cancellationToken"></param>
+        /// <returns>A task whose result (when complete) is the complete set of results fetched starting with the given
+        /// request, and continuing to make further requests until a response has no "next page" token.</returns>
+        public async Task<IList<TResource>> FetchAllAsync(
```

Comment:
I would say it **has to be List** here, to make it very clear that once the `Task` is `awaited`, you got all the results ready. If it returned `IEnumerable` that would not be clear. It could be that while enumerating the IEnumerable returned  by that function we may block yet again (just like in the `Fetch` now)